Pr 1 of 2: Add Microarray Pathway Analysis - GSVA example (#359)
AlexsLemonade/refinebio-examples: 02-microarray/pathway-analysis_microarray_03_gsva.Rmd

---
title: "Gene set variation analysis - Microarray"
author: "CCDL for ALSF"
date: "November 2020"
output:  
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
---

# Purpose of this analysis

This example is one of pathway analysis module set, we recommend looking at the [pathway analysis introduction](https://alexslemonade.github.io/refinebio-examples/02-microarray/pathway-analysis_microarray_00_intro.html) to help you determine which pathway analysis method is best suited for your purposes.

In this example we will cover a method called Gene Set Variation Analysis (GSVA) to calculate gene set or pathway scores on a per-sample basis.
Gene set variation analysis (GSVA) constitutes a starting point to build pathway-centric models of biology.
Rather than contextualizing some results you have received from another analysis like DGE, GSVA is designed to provide an estimate of pathway variation for each of the samples in an experiment.

> This means that GSVA scores will depend on the samples included in the dataset when you run GSVA; if you added more samples and reran GSVA, you would expect the scores to change [@Hanzelmann2013].

Reviewer: ```suggestion
This means that GSVA scores will depend on the samples included in the dataset when you run GSVA; if you added more samples and ran GSVA again, you would expect the scores to change [@Hanzelmann2013].
```

> GSVA determines the relative enrichment of gene sets across samples using
> a non-parametric approach.
> GSVA transforms a gene by sample gene expression matrix into a gene set by sample pathway enrichment matrix [@Hanzelmann-github].

Reviewer: ```suggestion
GSVA determines the relative enrichment of gene sets across samples using a non-parametric approach.
GSVA transforms a gene by sample gene expression matrix into a gene set by sample pathway enrichment matrix [@Hanzelmann-github].
```

  # Install this package if it isn't installed yet
  BiocManager::install("GSVA", update = FALSE)
}

if (!("qusage" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("qusage", update = FALSE)
}

if (!("org.Hs.eg.db" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("org.Hs.eg.db", update = FALSE)
}
```

Attach the packages we need for this analysis.

```{r}
# Attach the `qusage` library
library(qusage)

# Attach the `GSVA` library
library(GSVA)

# Human annotation package we'll use for gene identifier conversion
library(org.Hs.eg.db)

# We will need this so we can use the pipe: %>%
library(magrittr)
```

## Import and set up data

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read both TSV files and add them as data frames to your environment.

We stored our file paths as objects named `metadata_file` and `data_file` in [this previous step](#check-out-our-file-structure).

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(metadata_file)

# Read in data TSV file

> df <- readr::read_tsv(data_file) %>%

Reviewer: Can we make this more descriptive than `df`? This comment will end up propagating throughout.

  # Tuck away the gene ID  column as row names
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>%
  dplyr::select(metadata$geo_accession)

# Check if this is in the same order
all.equal(colnames(df), metadata$geo_accession)

# Bring back the "Gene" column in preparation for mapping
df <- df %>%
  tibble::rownames_to_column("Gene")
```

### Import Gene Sets

> The function that we will use to run GSVA wants the gene sets to be in a list, rather than a tidy data frame that we used with `clusterProfiler` (although it does accept other formats).

Reviewer: I think this wording might come from training materials. In this context, there's no guarantee that someone will have looked at the examples that use `clusterProfiler`, so I would recommend rephrasing this to not make reference to the other examples at all.

```{r}
# First let's create a mapped data frame we can join to the gene expression values
mapped_df <- data.frame(
  "entrez_id" = mapIds(
    org.Hs.eg.db, # Replace with annotation package for the organism relevant to your data
    keys = df$Gene,
    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID
  )
) %>%
  # If an Ensembl gene identifier doesn't map to a Entrez gene identifier, drop that
  # from the data frame
  dplyr::filter(!is.na(entrez_id)) %>%
  # Make an `Ensembl` column to store the row names
  tibble::rownames_to_column("Ensembl") %>%
  # Now let's join the rest of the expression data
  dplyr::inner_join(df, by = c("Ensembl" = "Gene"))
```

This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple Entrez IDs.
In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.

> For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.

Reviewer: ```suggestion
For the purpose of performing GSVA later in this notebook, we keep only the first mapped IDs.
```

In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.

Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.

```{r}
any(duplicated(mapped_df$entrez_id))
```

Looks like we do have duplicated Entrez IDs.
Let's find out which Entrez IDs have been duplicated.

```{r}
dup_entrez_ids <- mapped_df %>%
  dplyr::filter(duplicated(entrez_id)) %>%
  dplyr::pull(entrez_id)

dup_entrez_ids
```

#### Handling duplicate gene identifiers

As we mentioned earlier, we will not want any duplicate gene identifiers in our data frame when we convert it into a matrix in preparation for the GSVA steps later.
GSVA is executed on a per sample basis so let's keep the maximum expression value per sample associated with the duplicate Entrez gene identifiers.
In other words, we will keep only the maximum expression value found across the duplicate Entrez gene identifier instances for each sample or column in this case.

Let's take a look at the rows associated with the duplicated Entrez IDs and see how this will play out. 

```{r}
mapped_df %>%
  dplyr::filter(entrez_id %in% dup_entrez_ids)
```

> As an example using the strategy we described, for `GSM187153`'s data in the first column, `-0.007355525` is larger than `-0.136553967` so moving forward, Entrez gene `6013` will have ``-0.007355525` value and the `-0.136553967` would be dropped from the dataset. 

Reviewer: ```suggestion
As an example using the strategy we described, for `GSM187153`'s data in the first column, `-0.007355525` is larger than `-0.136553967` so moving forward, Entrez gene `6013` will have `-0.007355525` value and the `-0.136553967` would be dropped from the dataset. 
```

However, this is just one method of handling duplicate gene identifiers.
See the [Gene Set Enrichment Analysis (GSEA) User guide](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html) for more information on other expression values that can be used, like the median expression value for example [@GSEA-user-guide].

Now, let's implement our choose the max value method for all samples and Entrez IDs using tidyverse functions. 

```{r}
max_dup_df <- mapped_df %>%
  # We won't be using Ensembl IDs moving forward, so we will drop this column

>   dplyr::select(-"Ensembl") %>%

Reviewer: I don't think the quotes are necessary. It's also possible that the `group_by()` would have resulted in this column being dropped anyway, but I haven't tested this!

```suggestion
  dplyr::select(-Ensembl) %>%
```

  # Filter to include only the rows associated with the duplicate Entrez gene identifiers
  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
  # Group by Entrez IDs
  dplyr::group_by(entrez_id) %>%
  # Get the maximum expression value using all values associated with each duplicate
  # Entrez ID for each column or sample in this case
  dplyr::summarize_all(max)

max_dup_df
```
We can see `GSM187153` now has the `-0.007355525` value for Entrez ID `6013` like expected. 

Looks like we were able to successfully get rid of the duplicate Entrez gene identifiers!

Now let's combine our de-duplicated data with the rest of the mapped data!

```{r}
filtered_mapped_df <- mapped_df %>%
  # We won't be using Ensembl IDs moving forward, so we will drop this column

>   dplyr::select(-"Ensembl") %>%

Reviewer: Come to think of it, why not move this to the chunk that ends at line 307 so you don't have to do it twice? I guess a downside is that you wouldn't get to _show_ that different Ensembl IDs mapped to the same Entrez ID in lines 350-351 but you could _tell_.

For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.
For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).

Let's see a preview of `mapped_df`.

```{r}
head(mapped_df)
```

We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as rownames.
In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.

Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.

```{r}
any(duplicated(mapped_df$entrez_id))
```

Looks like we do have duplicated Entrez IDs.
Let's find out which Entrez IDs have been duplicated.

```{r}
dup_entrez_ids <- mapped_df %>%
  dplyr::filter(duplicated(entrez_id)) %>%
  dplyr::pull(entrez_id)

dup_entrez_ids
```

Now let's take a look at the rows associated with the duplicated Entrez IDs.

```{r}
mapped_df %>%
  dplyr::filter(entrez_id %in% dup_entrez_ids)
```

We can see that the associated values vary for each row.

> As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the maximum expression value per sample.

Reviewer 2: We may also want to remind the user that these values are representing probe IDs ultimately, even though the Ensembl column shows transcript IDs.  This is a tad different philosophically from RNA-seq where we've read the transcript itself; here we really are just dealing with probe differences here.

> As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the maximum expression value per sample.

Reviewer: Yes, but it's a little bit more nuanced that that because we're using BrainArray with SCAN. Quoting the [`SCAN.UPC`](https://bioconductor.org/packages/release/bioc/vignettes/SCAN.UPC/inst/doc/SCAN.vignette.pdf) documentation:

> By default,SCANmaps Affymetrix probes to “probeset” identifiers provided by themanufacturer.   However,  these  mappings  may  be  outdated  and  include  problematicprobes (for example, those that may cross hybridize).  In addition, multiple probesetsmay be assigned to a single gene.  As an alternative, the BrainArray resource (seehttp://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp) provides regularly updated mappings that exclude problematic probes and mapto genes rather than probesets.  

The target audience for this probably doesn't need all that information, so you can use terms like Ensembl gene identifier (not transcript identifiers, those would start with `ENST`).

> As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the maximum expression value per sample.

Reviewer 2: My use of the word transcript is confusing/vague/a bit wrong. I meant to differentiate it from rna-seq reads, but no, these are not transcript IDs.

See the [Gene Set Enrichment Analysis (GSEA) User guide](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html) for more information on other expression values that can be used [@GSEA-user-guide]. 

We are removing values for three gene identifiers here, so it is unlikely to have a considerable impact on our results.

```{r}
max_dup_df <- mapped_df %>%
  # Filter to include only the rows associated with the duplicate gene identifiers
  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
  # Group by Entrez IDs
  dplyr::group_by(entrez_id) %>%
  # Get the maximum expression value using all values associated with the grouping variable
  dplyr::summarize_all(max)

max_dup_df
```
Looks like we were able to successfully get rid of the duplicate gene identifiers and keep the observations with the maximum expression value!

Now let's combine the rest of the mapped data!

```{r}
filtered_mapped_df <- mapped_df %>%
  dplyr::filter(
    # First let's get the data associated with the gene identifiers that aren't duplicated
    !entrez_id %in% dup_entrez_ids |
      # Now let's get the data associated with the duplicated gene identifiers that we
      # collapsed in `max_dup_df`

>       entrez_id %in% max_dup_df$entrez_id &

Reviewer 2: As an example, I think this means you really are going to keep things based on what `max(c("ENSG00000107018", "ENSG00000107014"))` returns, which will just be based on character values of the id and not what you determined in `max_dup_df`.

>       entrez_id %in% max_dup_df$entrez_id &

Reviewer 2: I think you will want to keep the first part of your filter (remove duplicated IDs) and then bind on `max_dup_df`. May want to do some kind of arrange after that so the dups aren't all at the end or beginning.

>       entrez_id %in% max_dup_df$entrez_id &

Reviewer 2: I think to minimize some confusion, you can get rid of the `Ensembl` column earlier. Basically once your values are mixed up in `max_dup_df` the `Ensembl` column doesn't mean anything to us anymore anyway.

>       entrez_id %in% max_dup_df$entrez_id &

Author: Yes I see your point here. I could see how long it would take to run the steps on the whole mapped data frame rather than just on the data frame with the duplicated gene identifiers. That would be the ideal option here I believe (if not I'll try binding on `max_dup_df` again, it didn't work for me in development).

>       entrez_id %in% max_dup_df$entrez_id &

Reviewer 2: >  Run the steps on the whole mapped data frame rather than just on the data frame with the duplicated gene identifiers. That would be the ideal option here I believe. 

Can you explain why you think this is ideal? (What if a user has less memory and a much bigger dataset -- I think this idea may be unnecessarily memory using)

        Ensembl %in% max_dup_df$Ensembl
  )
```

Let's check to see that we removed the duplicate Entrez IDs from the rest of the mapped data.

```{r}
filtered_mapped_df %>%
  # Let's sort by the Entrez IDs we know were duplicated

>   dplyr::arrange(dplyr::desc(entrez_id %in% dup_entrez_ids)) %>%

Reviewer 2: If we are just doing a preview, I think this can be a lot simpler: 
```suggestion
  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
```

Author: I changed this to `head()` now that we arrange by the `dup_entrez_ids` in the previous step.

  # Now lt's display the first six observations of our new object
  head()
```

Great, looks like we were able to successfully get rid of the duplicate gene identifiers while joining the rest of the mapped data!

As mentioned earlier, we need a matrix for GSVA.
Let's now convert our data frame into a matrix and prepare our object for GSVA.

```{r}
filtered_mapped_matrix <- filtered_mapped_df %>%
  # First let's remove the column we no longer need

>   dplyr::select(-c("Ensembl")) %>%

Reviewer 2: I don't *think* you need this `c()`? 
```suggestion
  dplyr::select(-"Ensembl") %>%
```

We can see that the associated values vary for each row.

As we mentioned earlier, we will want to collapse any duplicate gene identifiers in preparation for the GSVA steps later.

> GSVA is executed on a per sample basis so let's keep the maximum expression value per sample associated with the duplicate gene identifiers.

Reviewer 2: ```suggestion
GSVA is executed on a per sample basis so let's keep the maximum expression value per sample associated with the duplicate Ensembl gene identifiers.
```

When there's multiple gene id types, it can help to remind us which one is the duplicate here that we are talking about.

Author: Actually, I believe the Entrez ID is the duplicate here, multiple Ensembl gene identifiers mapped to the same Entrez ID.

Reviewer 2: Ah right. I got confused too. We’ll all the more reason to specify.

> Note that using the maximum expression value to collapse duplicate gene identifiers means preferentially choosing a value or group of a higher rank downstream.

Reviewer 2: I'm not sure what this sentence means. Can explain what you mean here?

Author: This was based on [this comment from the draft PR](https://github.com/AlexsLemonade/refinebio-examples/pull/352#discussion_r524299264), from which I gathered that choosing the max expression value here would result in choosing the values of higher ranking in the downstream GSVA results. I can re-word this sentence for clarification.

Reviewer 2: I think “collapsing” identifiers in general is a phrase that makes sense, but I think once we’ve stated the idea, we can be more specific about what we are doing. You use the phrase “collapse identifiers” quite a few times back to back so I’m trying to think if there more specific, clear descriptions we can replace some of those repeats with.

Reviewer 2: This could also be a thing that I’m being weird about and the next reviewer might have a different idea or thoughts about it. 🤷‍♀️

Author: Let me know what you think of the latest changes in that case.

> However, this is just one of the expression values that can be used to collapse duplicate gene identifiers.

Reviewer 2: What is one of the values?

Author: I add an example in here.

max_dup_df
```
Looks like we were able to successfully collapse the duplicate gene identifiers by keeping the maximum expression value per sample associated with each of the duplicate gene identifiers!

Now let's combine the rest of the mapped data!

```{r}
filtered_mapped_df <- mapped_df %>%
  # First let's get the data associated with the gene identifiers that aren't duplicated
  dplyr::filter(!entrez_id %in% dup_entrez_ids) %>%
  # Now let's bind the rows of the data we collapsed in `max_dup_df`
  dplyr::bind_rows(max_dup_df) %>%
  # Arrange so that the Entrez IDs that were duplicated are at the top

>   dplyr::arrange(dplyr::desc(entrez_id %in% dup_entrez_ids)) %>%

Reviewer 2: Is there a reason we want the duplicates on the top? I think mixing in is fine? Or leaving as is?

Author: The reason behind this was in preparation of the next step, where we can use the `head()` function to show that we retained the collapsed duplicate data as well as the rest of the mapped data.

  # Now let's remove the column we no longer need
  dplyr::select(-"Ensembl")
```

Let's preview `filtered_mapped_df` to check that we combined the collapsed duplicate Entrez IDs data with the rest of the mapped data.

```{r}

> head(filtered_mapped_df)

Reviewer 2: If you are arranging early so the dups are at the top, we can leave the original `filtered_mapped_df`alone, but then here you can use `filtered_mapped_df %>% dplyr::filter(entrez_id %in% dup_entrez_ids)` to get the same effect.

Author: Gotcha 👍
